features/build.js

```diff
@@ -0,0 +1,360 @@
+createDirectory(outputDirectoryPath);
+
+const title = 'Ably';
+const subTitle = 'SDK Features';
+const documentWriter = new DocumentWriter(
+  { title: `${subTitle} | ${title}` },
+  fs.createWriteStream(path.join(outputDirectoryPath, 'index.html')),
+);
+
+documentWriter.document((contentWriter) => {
+  contentWriter.h(1, `${title} ${subTitle}`);
+
+  // Our convention, for this table, is borders to the right and bottom of cells.
+  // (only exception being the top row, where there's also a border to the top)
+  contentWriter.class('border-separate zero-border-spacing');
+
+  contentWriter.table((tableWriter) => {
+    renderTableHeaderRow(tableWriter, levelCount);
+
+    // Second Pass: Render rows.
+    generateTableRows(tableWriter, levelCount, [], object);
+  });
+});
+
+/**
+ * Render column headings to a table row.
+ *
+ * @param {TableWriter} writer The HTML table writer to use.
+ * @param {number} maximumLevel The maximum depth, previously measured.
+ */
+function renderTableHeaderRow(writer, maximumLevel) {
+  writer.class('align-top sticky top-0 bg-blue-700 text-white font-bold');
+  const commonCellStyle = 'pt-1 pb-2 border-y-4 border-white border-r-4 sticky top-0';
+  writer.row((rowWriter) => {
+    rowWriter.columnSpan(maximumLevel);
+    rowWriter.class(`pr-1 text-center ${commonCellStyle}`);
+    rowWriter.cell((cellContentWriter) => {
+      cellContentWriter.text('Feature');
+    });
+
+    // Specification Points
+    rowWriter.class(`px-1 ${commonCellStyle}`);
+    rowWriter.cell((cellContentWriter) => {
+      cellContentWriter.text('Specification');
+    });
+
+    // Conceptual Documentation Links and Synopsis
+    rowWriter.class(`px-1 ${commonCellStyle}`);
+    rowWriter.cell((cellContentWriter) => {
+      cellContentWriter.text('Synopsis and Links to Conceptual Documentation');
+    });
+
+    // API Documentation
+    rowWriter.class(`px-1 ${commonCellStyle}`);
+    rowWriter.cell((cellContentWriter) => {
+      cellContentWriter.text('Generic API Documentation');
+    });
+
+    // SDK columns
+    // eslint-disable-next-line no-restricted-syntax
+    for (const sdkManifestSuffix of sdkManifests.keys()) {
+      rowWriter.class(`px-1 text-center ${commonCellStyle}`);
+      rowWriter.cell((cellContentWriter) => {
+        cellContentWriter.text(sdkManifestSuffix);
+      });
+    }
+  });
+}
+
+/**
+ * Inspect a node, and it's children, optionally rendering to table rows.
+ *
+ * @param {TableWriter} [writer] The HTML table writer to use, or `null` to use this function purely to measure depth.
+ * @param {number} maximumLevel The maximum depth, previously measured or arbitrary.
+ * @param {string[]} parentKeys Parent keys, also indicating the depth of this node. Nodes at root have an empty array.
+ * @param {*} node The node.
+ * @returns {number} The number of levels, including this node and its children.
+ */
+function generateTableRows(writer, maximumLevel, parentKeys, node) {
```

Comment:
Kind of a long function, it even has region comments saying what the next bit does. Could we not turn those into well named private functions and replace comments with calls to same?

Reply:
It is, you're right. I am sure there could be opportunity to refactor this in due course, but for the moment it works as expected, the structure suits the stage of prototyping / incubation that I am at and it logically maps to the structure being output (the HTML output). Therefore I see it as a lower priority, for now, to reduce the line count for this single function simply because it's 'kind of long'. 😉 Were there to be a more specific, objective reason why a refactor would enhance maintainability then it might become more of a priority, but I believe the balance of comprehensibility is adequate.

Reply:
Fair enough, thank you for the background thinking.